server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -293,6 +301,14 @@ private ClusterQueryResult(Sequence<T> sequence, int numQueryServers)
       this.intervals = dataSourceAnalysis.getBaseQuerySegmentSpec()
                                          .map(QuerySegmentSpec::getIntervals)
                                          .orElseGet(() -> query.getIntervals());
+      this.cacheKeyManager = new CacheKeyManager<>(
```

Comment:
That is a good corner case. I didn't think of this. I don't see an easy flag to selectively disable result level caching for join. So I am just going to hard-code this in the new `CacheKeyManager`.  sounds good?

Reply:
Hmm, good catch. I think the real fix would probably be including the version of broadcast segments in the query which is sent to data servers.

Reply:
wouldn't it be better for data servers to send the segment version to the broker which it can either use in cache key or not cache at all. though doing this might be trickier than passing extra information in the query.

Reply:
I think @jihoonson's suggestion is best, because that would not only make caching correct, it would also ensure that the same broadcast table version is used across all data servers involved in a query. That'd make the query results more internally consistent (today, potentially, a different version of the table could be used on different data servers).

Reply:
I suppose, in that case, we also keep both the versions on a data server for a while. Or otherwise, the table will not be in a queryable state till all the data servers see the new version.

Reply:
Yeah, we'd need to do that. The Coordinator would need to participate in the logic somehow (it would need to _not_ issue any drops for the old segments until the new ones are fully loaded).